lrpar/src/lib/ctbuilder.rs

```diff
@@ -757,43 +771,45 @@ where
     fn rebuild_cache(&self, grm: &YaccGrammar<StorageT>) -> String {
         // We don't need to be particularly clever here: we just need to record the various things
         // that could change between builds.
-        let mut cache = String::new();
-        cache.push_str("\n/* CACHE INFORMATION\n");
-
+        //
         // Record the time that this version of lrpar was built. If the source code changes and
         // rustc forces a recompile, this will change this value, causing anything which depends on
         // this build of lrpar to be recompiled too.
         let build_time = env!("VERGEN_BUILD_TIMESTAMP");
         let grammar_path = self.grammar_path.as_ref().unwrap().to_string_lossy();
-        let mod_name = self.mod_name;
+        let mod_name = QuoteOption(self.mod_name);
         let recoverer = self.recoverer;
         let yacckind = self.yacckind;
-        let visibility = self.visibility.cow_str();
+        let visibility = self.visibility.clone();
         let error_on_conflicts = self.error_on_conflicts;
-        writeln!(cache, "   Build time: {}", quote!(#build_time)).ok();
-        writeln!(cache, "   Grammar path: {}", quote!(#grammar_path)).ok();
-        writeln!(cache, "   Mod name: {}", quote!(#mod_name)).ok();
-        writeln!(cache, "   Recoverer: {}", quote!(#recoverer)).ok();
-        writeln!(cache, "   YaccKind: {}", quote!(#yacckind)).ok();
-        writeln!(cache, "   Visibility: {}", quote!(#visibility)).ok();
-        writeln!(
-            cache,
-            "   Error on conflicts: {}\n",
-            quote!(#error_on_conflicts)
-        )
-        .ok();
-
-        // Record the rule IDs map
-        for tidx in grm.iter_tidxs() {
-            let n = match grm.token_name(tidx) {
-                Some(n) => format!("'{}'", n),
-                None => "<unknown>".to_string(),
-            };
-            writeln!(cache, "   {} {}", usize::from(tidx), n).ok();
-        }
 
-        cache.push_str("*/\n");
-        cache
+        let rule_map = grm
+            .iter_tidxs()
+            .map(|tidx| {
+                QuoteTuple((
+                    usize::from(tidx),
+                    grm.token_name(tidx).unwrap_or("<unknown>"),
+                ))
+            })
+            .collect::<Vec<_>>();
+        let rule_map_len = rule_map.len();
+        let cache_module = quote! {
+            #[allow(unused)]
+            mod _cache_information_ {
+                use ::lrpar::{RecoveryKind, Visibility};
+                use ::cfgrammar::yacc::YaccKind;
+
+                const BUILD_TIME: &'static str = #build_time;
```

Comment:
Both the old cache code, and this PR are missing the `RustEdition` variable,
I'll do that in a follow up patch if that is okay.

Reply:
Works for me.